### DISCUSSION TITLE
Eta reduction should works with implicit operator

### DISCUSSION THREAD

**Discussion by ghostnguyen** [dotnet/csharplang#8048] [General]
```cs
public class Foo<T>
{
    public T1 Process<T1>(Func<T1> fa, Func<T, T1> fb) => default;

    public static implicit operator Foo<T>(T t) => new();

    public Foo<T1> CallProcess1<T1>(Func<T, T1> f) => Process(() => new Foo<T1>(), v => f(v));
    public Foo<T1> CallProcess2<T1>(Func<T, T1> f) => Process(() => new Foo<T1>(), f); // compile error CS0411
}
```

The methods CallProcess1 and CallProcess2 should be identical but the CallProcess2 gets compile error CS0411: The type arguments for method cannot be inferred from the usage

But it works in this cases:

```cs
public class Foo2<T>
{
    public T1 Process<T1>(Func<T, T1> fb) => default;

    public static implicit operator Foo2<T>(T t) => new();

    public Foo2<T1> CallProcess1<T1>(Func<T, T1> f) => Process(v => f(v));
    public Foo2<T1> CallProcess2<T1>(Func<T, T1> f) => Process(f); // compile success
}
```

**Comment by colejohnson66** [comment]
Is this actually a language spec issue or is it just a Roslyn bug?

**Reply by ghostnguyen** [reply]
I think it's a Roslyn bug.

**Reply by huoyaoyuan** [reply]
It's usually not a compiler bug, but the language spec doesn't allow type inference like this.
C# generally tends to make the inference rules straight. It's helpful with complexity and compilation time.

**Reply by ghostnguyen** [reply]
But it's working in the second example.

**Reply by huoyaoyuan** [reply]
Lambdas and delegates are complex because of target-typing natural. I can't give any answer for this case.

**Comment by jilles-sg** [comment]
I think this is working as intended: `f` and `v => f(v)` are not the same. When writing `f`, the delegate is passed as is and only delegate variance is available (see e.g. https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/covariance-contravariance/variance-in-delegates ). When writing `v => f(v)`, a new delegate may be created and most implicit conversions are available for the argument and the return value.

Another way to see this is that `CallProcess2` in the first example must pass a delegate to `Process` that invokes both the original delegate and the conversion operator. Therefore, the delegate passed to `Process` cannot be the same as the original delegate.

Implicit conversion operators are not magic at run time. The compiler inserts calls to them, and at run time they are method calls like any other. (With `dynamic` the call might be inserted at run time.)

On the other hand, delegate variance works because implicit reference conversions do nothing at run time, so the same delegate object can be used.

As for the CS0411 error, try to specify the type arguments. There is no combination of type arguments that makes it work.

**Reply by ghostnguyen** [reply]
But both f and v => f(v) are delegate and same as covariance (or contravariance). That why it's work in the 2nd example.

So I think it's the issue of Roslyn override resolution

**Reply by jilles-sg** [reply]
The second example works because it applies the conversion operator to the return value of `Process` and can use the original delegate as is.

**Reply by ghostnguyen** [reply]
Then it should be apply the same to the 1st case without any issue.

**Reply by CyrusNajmabadi** [reply]
That would allocate a new delegate, violating the current expectation that you are just passing a specific value along.

**Reply by ghostnguyen** [reply]
But it works on the 2nd example

**Reply by colejohnson66** [reply]
The second example doesn’t invoke the implicit conversion

**Reply by ghostnguyen** [reply]
The 2nd example does invoke the implicit conversion. You can test by removing the implicit conversion.

**Reply by roald-di** [reply]
colejohnson66 is actually right, in the second example the implicit conversion is applied to the return value of `Process` and not to `f`.
That's because the type parameter of `Process` is inferred to be `T1` so no conversion for `f` is necessary.

If you change the second example to

```csharp
public Foo2<T1> CallProcess1<T1>(Func<T, T1> f) => Process<Foo2<T1>>(v => f(v));
public Foo2<T1> CallProcess2<T1>(Func<T, T1> f) => Process<Foo2<T1>>(f);
```

you will get the same behavior as in the first example.

**Reply by ghostnguyen** [reply]
But the main issue still there:
`f` and `v => f(v)` should be equal (`Func` is also a delegate type) and should be used interchangeable as eta reduction.

**Reply by colejohnson66** [reply]
They're not.

`f` means to "invoke `f` and *nothing else*." If the types aren't compatible, it won't compile. `v=>f(v)` means to "invoke any conversions necessary to turn `v` into something that can be passed to `f`, then invoke `f`, then invoke any conversions necessary to turn the result into something compatible with the expected return type".

You have to remember that casts are just normal method calls with fancy syntax. Specifically, `v => f(v)` gets "lowered" as: `v => (TResult)f((TInput)v)`. So, from your first example, `v => f(v)` gets lowered as `v => 'operator Foo<T>'(f(v))`. When the casts are made explicit, you can easily see how `f` and `v => f(v)` are different methods - because they're different code!

From your second example:

```cs
public Foo2<T1> CallProcess2<T1>(Func<T, T1> f) => Process(f);
```

That gets compiled as: `(Foo2<T1>)Process(f)`; You just don't see that.

**Reply by ghostnguyen** [reply]
I get how it works currently.

But as I suggest, it _should also work_ in more generic way. 

For example, 
1. Reduce/convert/normalize `Func<>` and `delegate` to the same kind of object if signature match.
2. Then apply the same procedures: Checking override, type match, etc.. and perform the same actions: invoke, implicit convert, etc..

**Reply by CyrusNajmabadi** [reply]
As I mentioned above: 

That would allocate a new delegate, violating the current expectation that you are just passing a specific value along.

**Reply by ghostnguyen** [reply]
The given value f indeeds a delegate type.
